awx_collection/plugins/module_utils/tower_api.py

```diff
@@ -567,6 +574,30 @@ def create_if_needed(self, existing_item, new_item, endpoint, on_create=None, it
         else:
             self.exit_json(**self.json_output)
 
+    # We need to be able to recursevly step through fields in the case of inputs for credentials.
+    # They are dicts and we can't just compare them at the top level because the dict from the tower api may have more fields that we have.
+    # For example, say someone did:
+    #  - tower_credential:
+    #      name: 'a cred'
+    #      username: 'John'
+    # Our new dict would be like { 'username': 'new_name' }
+    # But the existing cred from tower might come back as: { 'username': 'new_name', 'password': '$encrypted$', 'field2': 'something else' }
```

Comment:
I don't follow this. It's not that your argument here is wrong, but the logic here is written from the perspective of getting the API data and assessing whether a change _was_ made. That's not how it's used. This is used to determine whether a change _should be_ made, before the request is issued.

If I give a credential `inputs` field with the keys username & password, and the current API data has `inputs` of username & password & field2, then the PATCH _should_ be made in order to remove "field2". Otherwise there's no way to remove it through these modules.

A dictionary field provided to the module should be a _full specification_ of what data should be put in that field. It is still possible that the API validator _changes_ that data, but there is no reasonable way to know in advance of making the request what changes might be made. Thus, what I was saying in https://github.com/ansible/awx/issues/5474 that you could have 2 methods, one that determines if a change _should_ be made, and one that determines if a change _was actually_ made after the fact. Some people still might not like that design (it's complex and does unexpected things), but the API leaves no choice.

----
There's no way we want to fully solve that problem right now, with this conversion. If the API inserts "default" values into some fields unexpectedly, then I don't think this module should worry about a correct changed value in that case for now.

On the other hand, I can see the omission of comparing fields present server-side but not client-side causing near-terms bugs and regressions with this change.